x-pack/plugin/async-search/src/main/java/org/elasticsearch/xpack/search/AsyncSearchTask.java

```diff
@@ -284,15 +286,24 @@ private void executeCompletionListeners() {
             }
             hasCompleted = true;
         }
-        AsyncSearchResponse finalResponse = getResponse();
+        // we don't need to restore the response headers, they should be included in the current
+        // context since we are called by the search action listener.
+        AsyncSearchResponse finalResponse = getResponse(false);
         for (Consumer<AsyncSearchResponse> listener : completionListeners.values()) {
             listener.accept(finalResponse);
         }
         completionListeners.clear();
     }
 
-    private AsyncSearchResponse getResponse() {
+    /**
+     * Returns the current {@link AsyncSearchResponse} and restores the response headers
+     * in the local thread context depending on the value of <code>restoreResponseHeaders</code>.
+     */
+    private AsyncSearchResponse getResponse(boolean restoreResponseHeaders) {
```

Comment:
I am not a fan of boolean flags, this method looks simple enough that we could split in two: getResponse and getResponseWithHeaders or something along those lines?

Reply:
++, I pushed https://github.com/elastic/elasticsearch/pull/54349/commits/f6e8ab01a0564e371793d79a7d3dd145732783d0